docs/tutorials/signing/vuln-attestation.md

```diff
@@ -1,35 +1,137 @@
+
+For example:
+```
+cosign attest --predicate scan.json --type vuln docker.io/anaisurlichs/signed-example@sha256:c5911ac313e0be82a740bd726dc290e655800a9588424ba4e0558c705d1287fd 
 ```
 
 Note: Replace the container image with the container image that you would like to scan.
+Next, Sigstore will ask you to verify with an account -- Microsoft, GitHub, or Google.
+
+Once done, the user will be provided with a certificate in the terminal where they ran the command. This certificate can be used to verify the attestation in a later stage. Example certificate:
+```
+-----BEGIN CERTIFICATE-----
+MIIC1TCCAlygAwIBAgIUfSXI7xTWSLq4nuygd8YPuhPZlEswCgYIKoZIzj0EAwMw
+NzEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MR4wHAYDVQQDExVzaWdzdG9yZS1pbnRl
+cm1lZGlhdGUwHhcNMjQwMTExMTMzODUzWhcNMjQwMTExMTM0ODUzWjAAMFkwEwYH
+KoZIzj0CAQYIKoZIzj0DAQcDQgAETcUNnK76mfo9G3j1c7NN6Vcn6yQPDX5rd3QB
+unkHs1Uk59CWv3qm6sUyRNYaATs9zdHAZqLck8G4P/Pj7+GzCKOCAXswggF3MA4G
+........
+-----END CERTIFICATE-----
+```
+
+## Option 2: Signing and Generating an attestation with a new Cosign key pair
+
+To generate an attestation for the container image with a separate key pair, we can use Cosign to generate a new key pair:
+```
+cosign generate-key-pair 
+```
+
+This will generate a `cosign.key` and a `cosign.pub` file. The `cosign.key` file is your private key that should be kept confidential as it is used to sign artefacts. However, the `cosign.pub` file contains the information of the corresponding public key. This key can be used by third parties to verify the attestation -- basically that this person who claims to have signed the attestation actually is the one who signed it. 
+
+#### Signing
+
+Sign the container image:
+```
+cosign sign --key cosign.key docker.io/anaisurlichs/signed-example@sha256:c5911ac313e0be82a740bd726dc290e655800a9588424ba4e0558c705d1287fd
+```
+
+#### Attestation
+
+To generate the attestation with the specific key pairs, run the following command:
+```
+cosign attest --key cosign.key --type vuln --predicate scan.json docker.io/anaisurlichs/signed-example@sha256:c5911ac313e0be82a740bd726dc290e655800a9588424ba4e0558c705d1287fd 
+```
+
+## Verify the attestation
+
+### Option 1 -- No separate key pair
+
+If you have not generated a key pair but received a certificate after the container image was signed, use the following command to verify the attestation:
+
+```
+cosign verify-attestation --certificate=path-to-the-certificate --type vuln --certificate-identity Email-used-to-sign  --certificate-oidc-issuer='the-issuer-used' docker.io/DockerHubID/imagename:imageSHA
+```
+
+For example, the command could be like this:
+```
+cosign verify-attestation --certificate=cert.txt --type vuln --certificate-identity urlichsanais@gmail.com  --certificate-oidc-issuer='https://github.com/login/oauth' anaisurlichs/signed-example@sha256:c5911ac313e0be82a740bd726dc290e655800a9588424ba4e0558c705d1287fd
```

Comment:
I'm not sure if it's an issue with my local environment, but I've tried several times and the verify-attestation command with the --certificate option is not succeeding. Were you able to execute it successfully in your environment?

<details><summary>Details</summary>
<p>

```
> cosign version
  ______   ______        _______. __    _______ .__   __.
 /      | /  __  \      /       ||  |  /  _____||  \ |  |
|  ,----'|  |  |  |    |   (----`|  | |  |  __  |   \|  |
|  |     |  |  |  |     \   \    |  | |  | |_ | |  . `  |
|  `----.|  `--'  | .----)   |   |  | |  |__| | |  |\   |
 \______| \______/  |_______/    |__|  \______| |__| \__|
cosign: A tool for Container Signing, Verification and Storage in an OCI registry.

GitVersion:    v2.2.2
GitCommit:     unknown
GitTreeState:  unknown
BuildDate:     unknown
GoVersion:     go1.21.0
Compiler:      gc
Platform:      darwin/arm64

> trivy image --format cosign-vuln -o vuln.json otms61/my-alpine-2@sha256:ab6368870f017b2b33ff69a403fa756a087ab2b7914b5a50dc400193c7314df0
2024-01-17T09:05:52.376+0900	INFO	Need to update DB
2024-01-17T09:05:52.376+0900	INFO	DB Repository: ghcr.io/aquasecurity/trivy-db
2024-01-17T09:05:52.376+0900	INFO	Downloading DB...
42.31 MiB / 42.31 MiB [--------------------------------------------------------------------] 100.00% 30.34 MiB p/s 1.6s
2024-01-17T09:05:55.102+0900	INFO	Vulnerability scanning is enabled
2024-01-17T09:05:55.102+0900	INFO	Secret scanning is enabled
2024-01-17T09:05:55.102+0900	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2024-01-17T09:05:55.102+0900	INFO	Please see also https://aquasecurity.github.io/trivy/v0.46/docs/scanner/secret/#recommendation for faster secret detection
2024-01-17T09:05:55.119+0900	INFO	Detected OS: alpine
2024-01-17T09:05:55.119+0900	INFO	Detecting Alpine vulnerabilities...
2024-01-17T09:05:55.119+0900	INFO	Number of language-specific files: 0
2024-01-17T09:05:55.120+0900	WARN	This OS version is no longer supported by the distribution: alpine 3.15.0
2024-01-17T09:05:55.120+0900	WARN	The vulnerability detection may be insufficient because security updates are not provided

> cosign attest --type vuln --predicate vuln.json otms61/my-alpine-2@sha256:ab6368870f017b2b33ff69a403fa756a087ab2b7914b5a50dc400193c7314df0
Generating ephemeral keys...
Retrieving signed certificate...

	The sigstore service, hosted by sigstore a Series of LF Projects, LLC, is provided pursuant to the Hosted Project Tools Terms of Use, available at https://lfprojects.org/policies/hosted-project-tools-terms-of-use/.
	Note that if your submission includes personal data associated with this signed artifact, it will be part of an immutable record.
	This may include the email address associated with the account with which you authenticate your contractual Agreement.
	This information will be used for signing this artifact and will be stored in public transparency logs and cannot be removed later, and is subject to the Immutable Record notice at https://lfprojects.org/policies/hosted-project-tools-immutable-records/.

By typing 'y', you attest that (1) you are not submitting the personal data of any other person; and (2) you understand and agree to the statement and the Agreement terms at the URLs listed above.
Are you sure you would like to continue? [y/N] y
Your browser will now be opened to:
https://oauth2.sigstore.dev/auth/auth?access_type=online&client_id=sigstore&code_challenge=RPEMEDVLVi6Hv61KwpS1E8iSDiQ7SK2QQvPxEXhm1gI&code_challenge_method=S256&nonce=2b3iEyvxWlzxPbRZYymEJWoOTru&redirect_uri=http%3A%2F%2Flocalhost%3A61843%2Fauth%2Fcallback&response_type=code&scope=openid+email&state=2b3iF4wsjFvzyzf7gfYpUSWTYBs
Successfully verified SCT...
Using payload from: vuln.json
using ephemeral certificate:
-----BEGIN CERTIFICATE-----
MIIC1jCCAl2gAwIBAgIUElLm9+kPmvGmV11WTc/aOBsBNwAwCgYIKoZIzj0EAwMw
NzEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MR4wHAYDVQQDExVzaWdzdG9yZS1pbnRl
cm1lZGlhdGUwHhcNMjQwMTE3MDAwNzM5WhcNMjQwMTE3MDAxNzM5WjAAMFkwEwYH
KoZIzj0CAQYIKoZIzj0DAQcDQgAEqFfZBkOb9ar3+N8LVLJ395tt0urqSEyrS7ZG
XQMIHih7BrGXi+SH6EloBiZEgJB5YQiZan5KoRZuZsDMdHlkJaOCAXwwggF4MA4G
A1UdDwEB/wQEAwIHgDATBgNVHSUEDDAKBggrBgEFBQcDAzAdBgNVHQ4EFgQU4Y8C
8Yf3Vk1CFPpnE7bE5ZYRv38wHwYDVR0jBBgwFoAU39Ppz1YkEZb5qNjpKFWixi4Y
ZD8wJQYDVR0RAQH/BBswGYEXc2Fzb2FraXJhNjExNEBnbWFpbC5jb20wLAYKKwYB
BAGDvzABAQQeaHR0cHM6Ly9naXRodWIuY29tL2xvZ2luL29hdXRoMC4GCisGAQQB
g78wAQgEIAweaHR0cHM6Ly9naXRodWIuY29tL2xvZ2luL29hdXRoMIGLBgorBgEE
AdZ5AgQCBH0EewB5AHcA3T0wasbHETJjGR4cmWc3AqJKXrjePK3/h4pygC8p7o4A
AAGNFL61MAAABAMASDBGAiEAt/5GEaikQGwf8GWN9aUYw2KAmRYLffAn150fTfGH
qpQCIQD36MR0a3v0UOIHKakW5Xez70588DOz8E6fYmQ4jheaHjAKBggqhkjOPQQD
AwNnADBkAjA/cSj0+NFnfNTC5JsE0hy629r0ydEeyIgt2+wVU6OrkaSOjp87qois
pFFMxDREy50CMG7Xt2JAllHdp/LEwQLm2iAcMDYIEvGa/Tvy71eL19x8CnkRSZrX
+IoVq0ePtFXX1g==
-----END CERTIFICATE-----

tlog entry created with index: 64197639

> cat > cert.txt
-----BEGIN CERTIFICATE-----
MIIC1jCCAl2gAwIBAgIUElLm9+kPmvGmV11WTc/aOBsBNwAwCgYIKoZIzj0EAwMw
NzEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MR4wHAYDVQQDExVzaWdzdG9yZS1pbnRl
cm1lZGlhdGUwHhcNMjQwMTE3MDAwNzM5WhcNMjQwMTE3MDAxNzM5WjAAMFkwEwYH
KoZIzj0CAQYIKoZIzj0DAQcDQgAEqFfZBkOb9ar3+N8LVLJ395tt0urqSEyrS7ZG
XQMIHih7BrGXi+SH6EloBiZEgJB5YQiZan5KoRZuZsDMdHlkJaOCAXwwggF4MA4G
A1UdDwEB/wQEAwIHgDATBgNVHSUEDDAKBggrBgEFBQcDAzAdBgNVHQ4EFgQU4Y8C
8Yf3Vk1CFPpnE7bE5ZYRv38wHwYDVR0jBBgwFoAU39Ppz1YkEZb5qNjpKFWixi4Y
ZD8wJQYDVR0RAQH/BBswGYEXc2Fzb2FraXJhNjExNEBnbWFpbC5jb20wLAYKKwYB
BAGDvzABAQQeaHR0cHM6Ly9naXRodWIuY29tL2xvZ2luL29hdXRoMC4GCisGAQQB
g78wAQgEIAweaHR0cHM6Ly9naXRodWIuY29tL2xvZ2luL29hdXRoMIGLBgorBgEE
AdZ5AgQCBH0EewB5AHcA3T0wasbHETJjGR4cmWc3AqJKXrjePK3/h4pygC8p7o4A
AAGNFL61MAAABAMASDBGAiEAt/5GEaikQGwf8GWN9aUYw2KAmRYLffAn150fTfGH
qpQCIQD36MR0a3v0UOIHKakW5Xez70588DOz8E6fYmQ4jheaHjAKBggqhkjOPQQD
AwNnADBkAjA/cSj0+NFnfNTC5JsE0hy629r0ydEeyIgt2+wVU6OrkaSOjp87qois
pFFMxDREy50CMG7Xt2JAllHdp/LEwQLm2iAcMDYIEvGa/Tvy71eL19x8CnkRSZrX
+IoVq0ePtFXX1g==
-----END CERTIFICATE-----

> cosign verify-attestation --certificate=cert.txt --type vuln --certificate-identity sasoakira6114@gmail.com --certificate-oidc-issuer='https://github.com/login/oauth' otms61/my-alpine-2:latest
Error: no matching attestations: error verifying bundle: comparing public key PEMs, expected -----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEqFfZBkOb9ar3+N8LVLJ395tt0urq
SEyrS7ZGXQMIHih7BrGXi+SH6EloBiZEgJB5YQiZan5KoRZuZsDMdHlkJQ==
-----END PUBLIC KEY-----
, got -----BEGIN CERTIFICATE-----
MIIC1jCCAl2gAwIBAgIUElLm9+kPmvGmV11WTc/aOBsBNwAwCgYIKoZIzj0EAwMw
NzEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MR4wHAYDVQQDExVzaWdzdG9yZS1pbnRl
cm1lZGlhdGUwHhcNMjQwMTE3MDAwNzM5WhcNMjQwMTE3MDAxNzM5WjAAMFkwEwYH
KoZIzj0CAQYIKoZIzj0DAQcDQgAEqFfZBkOb9ar3+N8LVLJ395tt0urqSEyrS7ZG
XQMIHih7BrGXi+SH6EloBiZEgJB5YQiZan5KoRZuZsDMdHlkJaOCAXwwggF4MA4G
A1UdDwEB/wQEAwIHgDATBgNVHSUEDDAKBggrBgEFBQcDAzAdBgNVHQ4EFgQU4Y8C
8Yf3Vk1CFPpnE7bE5ZYRv38wHwYDVR0jBBgwFoAU39Ppz1YkEZb5qNjpKFWixi4Y
ZD8wJQYDVR0RAQH/BBswGYEXc2Fzb2FraXJhNjExNEBnbWFpbC5jb20wLAYKKwYB
BAGDvzABAQQeaHR0cHM6Ly9naXRodWIuY29tL2xvZ2luL29hdXRoMC4GCisGAQQB
g78wAQgEIAweaHR0cHM6Ly9naXRodWIuY29tL2xvZ2luL29hdXRoMIGLBgorBgEE
AdZ5AgQCBH0EewB5AHcA3T0wasbHETJjGR4cmWc3AqJKXrjePK3/h4pygC8p7o4A
AAGNFL61MAAABAMASDBGAiEAt/5GEaikQGwf8GWN9aUYw2KAmRYLffAn150fTfGH
qpQCIQD36MR0a3v0UOIHKakW5Xez70588DOz8E6fYmQ4jheaHjAKBggqhkjOPQQD
AwNnADBkAjA/cSj0+NFnfNTC5JsE0hy629r0ydEeyIgt2+wVU6OrkaSOjp87qois
pFFMxDREy50CMG7Xt2JAllHdp/LEwQLm2iAcMDYIEvGa/Tvy71eL19x8CnkRSZrX
+IoVq0ePtFXX1g==
-----END CERTIFICATE-----

main.go:74: error during command execution: no matching attestations: error verifying bundle: comparing public key PEMs, expected -----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEqFfZBkOb9ar3+N8LVLJ395tt0urq
SEyrS7ZGXQMIHih7BrGXi+SH6EloBiZEgJB5YQiZan5KoRZuZsDMdHlkJQ==
-----END PUBLIC KEY-----
, got -----BEGIN CERTIFICATE-----
MIIC1jCCAl2gAwIBAgIUElLm9+kPmvGmV11WTc/aOBsBNwAwCgYIKoZIzj0EAwMw
NzEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MR4wHAYDVQQDExVzaWdzdG9yZS1pbnRl
cm1lZGlhdGUwHhcNMjQwMTE3MDAwNzM5WhcNMjQwMTE3MDAxNzM5WjAAMFkwEwYH
KoZIzj0CAQYIKoZIzj0DAQcDQgAEqFfZBkOb9ar3+N8LVLJ395tt0urqSEyrS7ZG
XQMIHih7BrGXi+SH6EloBiZEgJB5YQiZan5KoRZuZsDMdHlkJaOCAXwwggF4MA4G
A1UdDwEB/wQEAwIHgDATBgNVHSUEDDAKBggrBgEFBQcDAzAdBgNVHQ4EFgQU4Y8C
8Yf3Vk1CFPpnE7bE5ZYRv38wHwYDVR0jBBgwFoAU39Ppz1YkEZb5qNjpKFWixi4Y
ZD8wJQYDVR0RAQH/BBswGYEXc2Fzb2FraXJhNjExNEBnbWFpbC5jb20wLAYKKwYB
BAGDvzABAQQeaHR0cHM6Ly9naXRodWIuY29tL2xvZ2luL29hdXRoMC4GCisGAQQB
g78wAQgEIAweaHR0cHM6Ly9naXRodWIuY29tL2xvZ2luL29hdXRoMIGLBgorBgEE
AdZ5AgQCBH0EewB5AHcA3T0wasbHETJjGR4cmWc3AqJKXrjePK3/h4pygC8p7o4A
AAGNFL61MAAABAMASDBGAiEAt/5GEaikQGwf8GWN9aUYw2KAmRYLffAn150fTfGH
qpQCIQD36MR0a3v0UOIHKakW5Xez70588DOz8E6fYmQ4jheaHjAKBggqhkjOPQQD
AwNnADBkAjA/cSj0+NFnfNTC5JsE0hy629r0ydEeyIgt2+wVU6OrkaSOjp87qois
pFFMxDREy50CMG7Xt2JAllHdp/LEwQLm2iAcMDYIEvGa/Tvy71eL19x8CnkRSZrX
+IoVq0ePtFXX1g==
-----END CERTIFICATE-----
```

</p>
</details>

Reply:
ah that is interesting, yes it worked for me -- will try again

Reply:
it worked for me but it seems like you don't need to specify the certificate path

Reply:
Thank you for your confirming! I see. It seems to be a problem with my local environment.
I agree that specifying only `--certificate-identity` and `--certificate-oidc-issuer` is sufficient.